pyro/poutine/trace_poutine.py

```diff
@@ -159,11 +170,11 @@ def __call__(self, *args, **kwargs):
         stores the arguments and return value of the function in special sites,
         and returns self.fn's return value
         """
-        self.msngr.trace = Trace(graph_type=self.msngr.graph_type)
-        self.msngr.trace.add_node("_INPUT",
-                                  name="_INPUT", type="args",
-                                  args=args, kwargs=kwargs)
-        ret = super(TraceHandler, self).__call__(*args, **kwargs)
+        with self.msngr:
```

Comment:
Interesting, does this fix a bug related to exception handling? I don't understand the difference between old and new versions.

Reply:
This uses `self.msngr.__enter__` to reset `self.msngr.trace`, then adds the `_INPUT` site after the reset happens, allowing `poutine.trace` to be used correctly as a context manager.